### PR TITLE
CLN update Lasso:Baseline solver for a better doc

### PR DESCRIPTION
**Pull request by mathurinm**
-  added a comment to mention that parameters defined in `parameters` can be accessed without being set in an init.
-  replaced z by w
- avoided computation of X.T @ X for Lipschitz constant (costly)

### CODE REVIEW

File: benchmarks/lasso/solvers/baseline.py [hunk 1]
```diff
@@ -5,33 +5,34 @@
...
-            z_hat -= 1 / self.L * grad
-            z_hat = self.st(z_hat, 1 / self.L * self.lmbd)
+                w_old = w.copy()
+            grad = self.X.T.dot(self.X.dot(w) - self.y)
+            w -= 1 / self.L * grad
+            w = self.st(w, 1 / self.L * self.lmbd)
             if self.use_acceleration:
-                z_hat = z_hat + (t_old - 1.) / t_new * (z_hat - z_old)
-        self.z_hat = z_hat
+                w = w + (t_old - 1.) / t_new * (w - w_old)
```
Comment (agramfort) [inline benchmarks/lasso/solvers/baseline.py]: you should use inplace operations as much as possible.

File: benchmarks/lasso/solvers/baseline.py [hunk 2]
```diff
@@ -5,33 +5,34 @@
...
+        self.w = w
 
-    def st(self, z, mu):
-        return np.sign(z) * np.maximum(0, np.abs(z) - mu)
+    def st(self, w, mu):
+        return np.sign(w) * np.maximum(0, np.abs(w) - mu)
```
Comment (agramfort) [inline benchmarks/lasso/solvers/baseline.py]: st could operate inplace to avoid unecessary memory allocations

File: doc/how.rst [hunk 3]
```diff
@@ -61,7 +61,7 @@ from the union of the data and the objective parameters.
 
 You can define your custom solver.
 
-Example of hand written solver
+Example of hand written and parametrized solver
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (agramfort) [inline doc/how.rst]: missing ~~~~

File: benchmarks/lasso/solvers/baseline.py [hunk 4]
```diff
@@ -5,33 +5,34 @@
...
+    name = 'Baseline'  # proximal gradient, eventually accelerated
 
+    # any parameter defined here is accessible as a class attribute
     parameters = {'use_acceleration': [False, True]}
 
     def set_objective(self, X, y, lmbd):
         self.X, self.y, self.lmbd = X, y, lmbd
 
-        self.L = np.linalg.norm(self.X.dot(self.X.T), ord=2)
+        self.L = np.linalg.norm(self.X, ord=2) ** 2
```
Comment (mathurinm) [inline benchmarks/lasso/solvers/baseline.py]: @agramfort IMO this is not fair that the constant is precomputed, this should be done in `run()`. what's your take?

**Comment by agramfort** [comment]
I agree

>

File: benchmarks/lasso/solvers/baseline.py (hunk 1 above)
Comment (mathurinm) [reply benchmarks/lasso/solvers/baseline.py]: done

File: benchmarks/lasso/solvers/baseline.py (hunk 2 above)
Comment (mathurinm) [reply benchmarks/lasso/solvers/baseline.py]: I thought of using `shrink = np.maximum(0, 1 - mu / np.abs(w))` but this causes division by 0 and working with masks causes creation of other arrays.

In hindsight the gradient should be the larger cost of the iteration, have you observed the contrary in the past?

Comment (agramfort) [reply benchmarks/lasso/solvers/baseline.py]: i doub't know how much impact this will have true. But here you make 4 copies of w in one line

Comment (mathurinm) [reply benchmarks/lasso/solvers/baseline.py]:
```In [5]: a = np.array([0., 1.])                                                  

In [6]: a *= np.maximum(0, 1 - 0.5 / np.abs(a))                                 
/home/mathurin/miniconda3/bin/ipython:1: RuntimeWarning: divide by zero encountered in true_divide
  #!/home/mathurin/miniconda3/bin/python

In [7]: a                                                                       
Out[7]: array([0. , 0.5])
```
but this raises a warning bc of mu / 0

I think hopefully numpy reuses temp arrays now and only 2 arrays are needed

File: benchmarks/lasso/solvers/baseline.py [hunk 5]
```diff
@@ -5,33 +5,34 @@
...
         self.X, self.y, self.lmbd = X, y, lmbd
 
-        self.L = np.linalg.norm(self.X.dot(self.X.T), ord=2)
-
     def run(self, n_iter):
+        L = np.linalg.norm(self.X, ord=2) ** 2
+
         n_features = self.X.shape[1]
-        z_hat = np.zeros(n_features)
+        w = np.zeros(n_features)
```
Comment (tomMoral) [inline benchmarks/lasso/solvers/baseline.py]: If we decide to go that way I would say that this need to be  changed globally in the benchmark. I think I used `z_hat` everywhere but we can do this in a separate PR.

File: benchmarks/lasso/solvers/baseline.py (hunk 2 above)
Comment (tomMoral) [reply benchmarks/lasso/solvers/baseline.py]: Inplace soft thresholding:
```python
def st(w, mu):
    a -= np.clip(w, -mu, mu)
    return a
```

File: benchmarks/lasso/solvers/baseline.py (hunk 5 above)
Comment (mathurinm) [reply benchmarks/lasso/solvers/baseline.py]: +1 to merge this one and do another PR

**Comment by tomMoral** [comment]
Thx!
